chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/onetime/measures/AddMissingMeasuresToListingsJob.java

```diff
@@ -0,0 +1,199 @@
+import gov.healthit.chpl.listing.measure.ListingMeasureDAO;
+import gov.healthit.chpl.listing.measure.MeasureDAO;
+import gov.healthit.chpl.scheduler.job.CertifiedProduct2015Gatherer;
+import gov.healthit.chpl.service.CertificationCriterionService;
+import gov.healthit.chpl.service.CertificationCriterionService.Criteria2015;
+import gov.healthit.chpl.util.Util;
+import lombok.extern.log4j.Log4j2;
+
+@Log4j2(topic = "addMissingMeasuresToListingsJobLogger")
+public class AddMissingMeasuresToListingsJob extends CertifiedProduct2015Gatherer implements Job {
+
+    @Autowired
+    private MacraMeasureDAO macraMeasureDao;
+
+    @Autowired
+    private MeasureDAO measureDao;
+
+    @Autowired
+    private ListingMeasureDAO listingMeasureDao;
+
+    @Autowired
+    private CertificationCriterionService criterionService;
+
+    @Autowired
+    private JpaTransactionManager txManager;
+
+    @Value("${executorThreadCountForQuartzJobs}")
+    private Integer threadCount;
+
+    public AddMissingMeasuresToListingsJob() throws Exception {
+        super();
+    }
+
+    @Override
+    public void execute(JobExecutionContext jobContext) throws JobExecutionException {
+        SpringBeanAutowiringSupport.processInjectionBasedOnCurrentContext(this);
+        LOGGER.info("********* Starting the Add Missing Measures to Listings job. *********");
+
+        // We need to manually create a transaction in this case because of how AOP works. When a method is
+        // annotated with @Transactional, the transaction wrapper is only added if the object's proxy is called.
+        // The object's proxy is not called when the method is called from within this class. The object's proxy
+        // is called when the method is public and is called from a different object.
+        // https://stackoverflow.com/questions/3037006/starting-new-transaction-in-spring-bean
+        TransactionTemplate txTemplate = new TransactionTemplate(txManager);
+        txTemplate.setPropagationBehavior(TransactionDefinition.PROPAGATION_REQUIRES_NEW);
+        txTemplate.execute(new TransactionCallbackWithoutResult() {
+
+            @Override
+            protected void doInTransactionWithoutResult(TransactionStatus status) {
+                try {
+
+                    getAll2015CertifiedProducts(LOGGER, threadCount, false).stream()
+                        .forEach(listing -> addMissingMeasures(listing));
+
+                } catch (Exception e) {
+                    LOGGER.error(e);
+                }
+            }
+        });
+        LOGGER.info("********* Completed the Add Missing Measures to Listings job. *********");
+    }
+
+    private void addMissingMeasures(CertifiedProductSearchDetails listing) {
+        LOGGER.info("Adding missing measures to listing " + listing.getId());
+        addMissingMeasure(listing, Criteria2015.B_1_OLD, "RT7 EH/CAH Medicare and Medicaid PI", "RT7 EH/CAH Medicare PI");
+        replaceListingMeasuresWithUpdatedMeasures(listing);
```

Comment:
This is a very confusing method. Does this line need to be repeated? Why? Also, why isn't it repeated at the end of the method after doing the second g10 measure?

Reply:
Well each time I am adding replacement measures or possibly adding associated criteria to an existing replacement measure I wanted to set the new listing measures on the details object so that for the next criteria I'm working with the latest listing details. I don't know if any of them could overlap in this way but if I just add measure, add measure, add measure, add criteria to measure, add measure, etc and between all of those adding of things I don't go get the latest "current state" then it feels like I could/would be making changes to data that is stale. So I wanted the logic to go 1) add measure to listing, 2) update the details object i'm working from to have the latest measures (including the one i just added), 3) add next measure to listing 4) update the details object to have the latest measures 5) add next measure to listing, and so on and so forth. Open to method naming suggestions. We don't need this after the last g10 is done because there are no more updates to make.

Reply:
Okay. I can understand that reasoning. I might want to think on this more if it weren't a one-time job that we're removing after it runs. That said, it isn't, so I'm good with it